junit-plugin/src/main/scala/org/scalajs/junit/plugin/ScalaJSJUnitPlugin.scala

```diff
@@ -0,0 +1,522 @@
+ *               annotations = List(new Test)),
+ *           new JUnitMethodMetadata(name = "baz", id = "2",
+ *               annotations = List(new Test, new Ignore("baz not implemented yet")))
+ *       ),
+ *       moduleMethods(
+ *           new JUnitMethodMetadata(name = "beforeClass", id = "0",
+ *               annotations = List(new BeforeClass)))
+ *     )
+ *   }
+ *   def newInstance(): ScalaJSJUnitTest = new Foo()
+ *   def nvoke(methodId: String): Unit = {
+ *     if (methodId == "0") Foo.beforeClass()
+ *     else throw new NoSuchMethodException(methodId)
+ *   }
+ * }
+ * }}}
+ * The test framework will identify `Foo\$scalajs\$junit\$bootstrapper` as a test module
+ * because it extends `JUnitTestBootstrapper`. It will know which methods to run based
+ * on the info returned by Foo\$scalajs\$junit\$bootstrapper.metadata,
+ * it will create new test instances using `Foo\$scalajs\$junit\$bootstrapper.newInstance()`
+ * and it will invoke test methods using `scalajs\$junit\$invoke` methods on the
+ * module or instance.
+ */
+class ScalaJSJUnitPlugin(val global: Global) extends NscPlugin {
+
+  val name: String = "Scala.js JUnit plugin"
+
+  val components: List[NscPluginComponent] =
+    List(ScalaJSJUnitPluginComponent)
+
+  val description: String = "Makes JUnit test classes invokable in Scala.js"
+
+  object ScalaJSJUnitPluginComponent
+    extends plugins.PluginComponent with transform.Transform {
+
+    protected def newTransformer(
+        unit: global.CompilationUnit): global.Transformer = {
+      new ScalaJSJUnitPluginTransformer
+    }
+
+    val global: Global = ScalaJSJUnitPlugin.this.global
+    val phaseName: String = "junit-inject"
+    val runsAfter: List[String] = List("mixin")
+    override val runsBefore: List[String] = List("jscode")
+
+    class ScalaJSJUnitPluginTransformer extends global.Transformer {
+
+      import global._
+      import rootMirror.getRequiredClass
+
+      private val testAnnotationSymbol =
+        getRequiredClass("org.junit.Test")
+
+      private val fixMethodOrderAnnotationSymbol =
+        getRequiredClass("org.junit.FixMethodOrder")
+
+      private val annotationWhiteList = List(
+        testAnnotationSymbol,
+        getRequiredClass("org.junit.Before"),
+        getRequiredClass("org.junit.After"),
+        getRequiredClass("org.junit.BeforeClass"),
+        getRequiredClass("org.junit.AfterClass"),
+        getRequiredClass("org.junit.Ignore")
+      )
+
+      private val scalaJSJUnitTestType =
+        getRequiredClass("org.scalajs.junit.ScalaJSJUnitTest").toType
+
+      private val jUnitClassMetadataType =
+        getRequiredClass("org.scalajs.junit.JUnitClassMetadata").toType
+
+      private val jUnitTestMetadataType =
+        getRequiredClass("org.scalajs.junit.JUnitTestBootstrapper").toType
+
+      private val jUnitMethodMetadataTypeTree =
+        TypeTree(getRequiredClass("org.scalajs.junit.JUnitMethodMetadata").toType)
```

Comment:
Not those, probably not. They are only used to put into `TypeTree`s and into `info`s, not for *identification*.